File: leetCode/Python/20-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        # second time
        if not s or len(s) % 2 != 0:
            return False
        parentheses = {')':'(', '}':'{', ']':'['}
        seen = []
        for char in s:
            if char in parentheses.values():
                seen.append(char)
            elif char in parentheses.keys():
                if seen == [] or parentheses[char] != seen[-1]:
                    return False
                seen.pop()
            else:
                return False
        return len(seen) == 0


    

File: leetCode/JavaScript/src/questions/20-valid-parentheses/20-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
// https://leetcode.com/problems/valid-parentheses/
const isValid = function (s) {
  if (s == undefined || s.length % 2 !== 0) {
    return false;
  }
  const characters = { ')': '(', '}': '{', ']': '[' };
  const stack = [];

  for (let i = 0; i < s.length; i++) {
    const char = s.charAt(i);
    switch (char) {
      case '(':
        stack.push(char);
        break;
      case '[':
        stack.push(char);
        break;
      case '{':
        stack.push(char);
        break;
      default:
        if (characters[char] !== stack.pop()) {
          return false;
        }
    }
  }
  return stack.length == 0;
};
